robot/impl/resource_manager.go

```diff
@@ -1108,20 +1139,35 @@ func (manager *resourceManager) processResource(
 	return newRes, true, nil
 }
 
-// markResourceForUpdate marks the given resource in the graph to be updated. If it does not exist, a new node
-// is inserted. If it does exist, it's properly marked. Once this is done, all information needed to build/reconfigure
-// will be available when we call completeConfig.
-func (manager *resourceManager) markResourceForUpdate(name resource.Name, conf resource.Config, deps []string, revision string) error {
+// addToBeConstructedResource adds a new, unconfigured graph node for a resource to the
+// resource graph. If a resource with the given name already exists in the resource graph,
+// an error is returned. If that pre-existing resource is remote, it is also marked for
+// update to use the new config, as we prefer local resources to remotes in the event of
+// collision. If that pre-existing resource is local, the pre-existing resource is also
+// marked for removal, and no new graph node is created.
+func (manager *resourceManager) addToBeConstructedResource(
+	name resource.Name,
+	conf resource.Config,
+	deps []string,
+	revision string,
+) error {
 	gNode, hasNode := manager.resources.Node(name)
 	if hasNode {
-		gNode.SetNewConfig(conf, deps)
-		gNode.UpdatePendingRevision(revision)
-		// reset parentage
-		for _, parent := range manager.resources.GetAllParentsOf(name) {
-			manager.resources.RemoveChild(name, parent)
-		}
+		manager.logger.Errorw("Duplicate resource found; rename the resource or use `prefix` to disambiguate if either resource is remote",
+			"colliding name", name.String())
+		// Treat unknownModel as a sign that the existing resource is remote.
+		if gNode.ResourceModel() == unknownModel {
+			manager.logger.Errorw("Remote resource with colliding name will be replaced with local resource",
+				"colliding name", name.String())
+			return manager.markResourceForUpdate(name, conf, deps, revision)
+		}
+
+		manager.logger.Errorw("Neither resource with colliding name will be reachable through this machine until collision is fixed",
+			"colliding name", name.String())
+		manager.markResourcesRemoved([]resource.Name{name}, nil)
```

Comment:
_sigh_ @cheukt I'm running into what you're trying to solve with https://github.com/viamrobotics/rdk/pull/5113 with this change. In particular, we were relying on being able to re-add a modular resource to the graph on module reconfigure/rename/crash/reload and it just happily take the place of the old modular resource. Now, we get a duplicate resource error above and the original resource is marked for removal.

It'd be awesome to merge #5113 before this PR, so I can work on top of whatever system you land on.